espressostreamer/espresso_streamer.go

```diff
@@ -0,0 +1,306 @@
+	for _, url := range hotshotUrls {
+		client := espressoClient.NewClient(url)
+		if client == nil {
+			// we should be able to initialize
+			// all the clients
+			log.Crit("Failed to create espresso client", "url", url)
+			return nil
+		}
+		espressoClients = append(espressoClients, *client)
+	}
+
+	espressoTEEVerifierAbi, err := bridgegen.IEspressoTEEVerifierMetaData.GetAbi()
+	if err != nil {
+		log.Crit("Unable to find EspressoTEEVerifier ABI")
+	}
+	l1Client, err := ethclient.Dial(parentChainNodeUrl)
+	if err != nil {
+		log.Crit("Failed to create l1 client", "url", parentChainNodeUrl)
+		return nil
+	}
+
+	arbSys, err := precompilesgen.NewArbSys(types.ArbSysAddress, l1Client)
+	if err != nil {
+		log.Crit("Failed to create arbsys", "err", err)
+		return nil
+	}
+
+	// we initialze a l1 reader that will poll for header every 60 seconds
+	l1Reader, err := headerreader.New(context.Background(), l1Client, func() *headerreader.Config {
+		return &headerReaderConfig
+	}, arbSys)
+
+	if err != nil {
+		log.Crit("Failed to create l1 reader", "err", err)
+		return nil
+	}
+	return &EspressoStreamer{
+		espressoClients:               espressoClients,
+		nextHotshotBlockNum:           nextHotshotBlockNum,
+		retryTime:                     retryTime,
+		pollingHotshotPollingInterval: pollingHotshotPollingInterval,
+		namespace:                     namespace,
+		l1Reader:                      l1Reader,
+		espressoTEEVerifierABI:        espressoTEEVerifierAbi,
+		espressoTEEVerifierAddr:       espressoTEEVerifierAddress,
+		// Initally, we set the confirmed message pos to max uint64
+		// because we dont know the confirmed message pos yet
+		confirmedMessagePos: abi.MaxUint256.Uint64(),
+	}
+}
+
+func (s *EspressoStreamer) Refresh(ctx context.Context, fetchLastMessageAndEspressoBlock func() (uint64, uint64, error)) (bool, error) {
+	lastMessagePos, lastEspressoBlock, err := fetchLastMessageAndEspressoBlock()
+	if err != nil {
+		return false, err
+	}
+
+	// If the last message pos only increase by 1, we dont need to refresh
+	// because this is expected behavior.
+	if lastMessagePos == s.confirmedMessagePos+1 {
+		s.confirmedMessagePos = lastMessagePos
+		s.currentMessagePos = lastMessagePos + 1
+		return false, nil
+	}
+
+	// If last message pos didnt increase by and if not equal to the confirmed message pos
+	// something went wrong
+	// we reset the values to the values returned from the source of truth.
+	if lastMessagePos != s.confirmedMessagePos {
+		s.confirmedMessagePos = lastMessagePos
+		s.currentMessagePos = lastMessagePos + 1
+		s.nextHotshotBlockNum = lastEspressoBlock
+		s.messageWithMetadataAndPos = []*MessageWithMetadataAndPos{}
+		log.Info("Refreshing espresso streamer", "lastMessagePos", lastMessagePos, "lastEspressoBlock", lastEspressoBlock, "confirmedMessagePos", s.confirmedMessagePos)
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (s *EspressoStreamer) Next() (*MessageWithMetadataAndPos, error) {
```

Comment:
This handles skipping forward in messages if there has been a batch posted, but because next does not read from the l1 it is possible that we have incorrect behavior. Consider a case where we have called refresh, and then next after a reorg before queue messages from hotshot has been polled.

If this were to occur, we would set the next call to queueMessagesFromHotshot to rebuild the streamers buffer, but we would have an empty buffer until the next time it is polled. leading to unexpected instances of getting an error from Next().

Reply:
`Consider a case where we have called refresh, and then next after a reorg before queue messages from hotshot has been polled.`

Reorg possibility should not be there because we should only look at batch position which has been `finalized`. Can you think of any other case?